src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java

```diff
@@ -152,39 +152,44 @@ public long add(ByteBuffer term, T key, InvalidVectorBehavior behavior)
             validateIndexable(vector, similarityFunction);
         }
 
-        var bytesUsed = new AtomicLong();
-        var newVector = new AtomicBoolean();
+        var bytesUsed = 0L;
+        VectorPostings<T> postings = postingsMap.get(vector);
         // if the vector is already in the graph, all that happens is that the postings list is updated
         // otherwise, we add the vector in this order:
-        // 1. to the vectorValues
-        // 2. to the postingsMap
+        // 1. to the postingsMap
+        // 2. to the vectorValues
         // 3. to the graph
         // This way, concurrent searches of the graph won't see the vector until it's visible
         // in the other structures as well.
-        var postings = postingsMap.computeIfAbsent(vector, v -> {
-            var bytes = RamEstimation.concurrentHashMapRamUsed(1); // the new posting Map entry
-            var ordinal = nextOrdinal.getAndIncrement();
-            bytes += (vectorValues instanceof ConcurrentVectorValues)
-                     ? ((ConcurrentVectorValues) vectorValues).add(ordinal, vector)
-                     : ((CompactionVectorValues) vectorValues).add(ordinal, term);
-            bytes += VectorPostings.emptyBytesUsed() + VectorPostings.bytesPerPosting();
-            bytesUsed.addAndGet(bytes);
-            newVector.set(true);
-            var vp = new VectorPostings<T>(ordinal, key);
-            postingsByOrdinal.put(ordinal, vp);
-            return vp;
-        });
-        if (newVector.get())
+        if (postings == null)
         {
-            // add to the index outside of the computeIfAbsent block
-            bytesUsed.addAndGet(builder.addGraphNode(postings.getOrdinal(), vectorValues));
+            postings = new VectorPostings<T>(key);
+            if (postingsMap.putIfAbsent(vector, postings) == null)
+            {
+                // we won the race to add the new entry; assign it an ordinal and add to the other structures
+                var ordinal = nextOrdinal.getAndIncrement();
+                postings.setOrdinal(ordinal);
```

Comment:
One note is that this updates mutable state without explicitly ensuring safe publication. I haven't read through the whole write to heap then flush to disk logic, but if we had a write in one thread and then start flushing to disk in another, it seems plausible that there won't be anything forcing safe publication, which could lead to the `getOrdinal` returning a -1. It's probably pretty unlikely, but what I'm most concerned about is that if it only happens when we disable assertions, it'll be particularly tricky.

I think it might make sense to make `ordinal` a volatile, but I am also fine with deferring for now.

Reply:
Good catch, you're right.